### ISSUE TITLE
StackTracePreserveTests test fails

### ISSUE THREAD

**Issue by CarolEidt** [dotnet/coreclr#20322]
This test: baseservices\exceptions\StackTracePreserve\StackTracePreserveTests\StackTracePreserveTests.cmd

When run with `COMPlus_TieredCompilation=0` on Windows x64.

This passed with https://github.com/dotnet/coreclr/commit/8e880c667e999aa7d9da8d7ea30c2db92e419ba0
in this test run: https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/jitstress/job/x64_checked_windows_nt_no_tiered_compilation/76/

It failed with https://github.com/dotnet/coreclr/commit/645faa5b24c32741944cad55f58fd66d6c8abfa4 in this test run: https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/jitstress/job/x64_checked_windows_nt_no_tiered_compilation/77/

There were 2 intervening commits, one of which was "Typos", so presumably it was the update of the buildtools (#20286).

**Comment by CarolEidt** [comment]
@ViktorHofer

**Comment by CarolEidt** [comment]
This test is also failing on Ubuntu x64, and appears to have started failing at about the same time, so presumably also due to the build tools update.

**Comment by BruceForstall** [comment]
This also fails in the Ubuntu arm64 tests

**Comment by sandreenko** [comment]
@RussKeldorph is it a codegen issue? Should it be in Future milestone?

**Comment by RussKeldorph** [comment]
@sandreenko I should have written down my reasoning for triaging to `area-ExceptionHandling` because now I don't recall ever touching this issue.  My best guess is that I assumed it was a bad test case that is incompatible with tail calls since it matches stack frames, but I don't know for sure.

Would you mind looking into this?  If it is a codegen issue, please fix the area and set Milestone to 3.0.  If it's a test issue, I would add the `test issue` tag and let @sergiy-k triage it.

**Comment by RussKeldorph** [comment]
@sandreenko Actually, it's more likely I just wanted @sergiy-k to do the initial investigation since the test is in his area.  If you want to investigate, feel free.

**Comment by sandreenko** [comment]
The test failed in outerloop on some platforms (including windows where I tested it locally), so probably it is an unstable one.

**Comment by sandreenko** [comment]
> @sandreenko I should have written down my reasoning for triaging to area-ExceptionHandling because now I don't recall ever touching this issue. My best guess is that I assumed it was a bad test case that is incompatible with tail calls since it matches stack frames, but I don't know for sure.

You are right. On x64 `Scenario9HelperInner` marks `s_EDI.Throw();` as a profitable tail call, so when the exception happens we do not have `Scenario9HelperInner` on the stack.

**Comment by RussKeldorph** [comment]
@sandreenko Thanks!  @sergiy-k FYI.

**Comment by sergiy-k** [comment]
@janvorli, should we just delete this test or do you see value in it?

**Comment by janvorli** [comment]
I still see a value in these tests as a mean to verify that we generate correct stack traces. I would try to stabilize this test by removing its scenarios that are unstable or making them resilient to the unstability. There are 8 scenarios that it tests and from what @sandreenko said, it looks like only the last one is failing.

**Comment by sandreenko** [comment]
> it looks like only the last one is failing.

That is correct. However, it doesn't mean that in the future other cases won't fail (for example when we add tail call support for more targets),

You could mark more methods (or their proxy callers) that you expect on the stack with `NoOptimization` attribute, (we need both NoInlining and NoTailCalling, that doesn't exist in MethodImplOptions  enum), that will fix the issue.

**Comment by AndyAyersMS** [comment]
`NoInlining` implies no tail calling (either as callee or caller) -- we assume that `NoInlining` means you want to see the method in its own unique stack frame.

That's why it can be tricky to come up with repro cases for tail call issues -- you have to be careful about using `NoInlining`.

**Comment by sandreenko** [comment]
> NoInlining implies no tail calling (either as callee or caller) -- we assume that NoInlining means you want to see the method in its own unique stack frame.

Thanks, Andy, that is useful to know.
